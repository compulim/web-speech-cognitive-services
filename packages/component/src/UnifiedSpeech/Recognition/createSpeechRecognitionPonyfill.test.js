jest.useFakeTimers();

const MOCK_SPEECH_SDK = {
  AudioConfig: {
    fromDefaultMicrophoneInput: () => ({})
  },
  OutputFormat: {
    Detailed: 'DETAILED'
  },
  SpeechConfig: {
    fromSubscription: (subscriptionKey, region) => {
      return {
        region,
        subscriptionKey
      };
    }
  },
  SpeechRecognizer: class {
    constructor() {
      this.canceled = this.recognized = this.recognizing = () => {};
    }
  }
};

const SPEECH_EVENTS = [
  'audioend',
  'audiostart',
  'end',
  'error',
  'result',
  'soundend',
  'soundstart',
  'speechend',
  'speechstart',
  'start'
];

function captureSpeechEvents(speech) {
  const queue = [];

  for (let eventName of SPEECH_EVENTS) {
    speech.addEventListener(eventName, event => queue.push(event));
  }

  return () => [...queue];
}

beforeEach(() => {
  jest.resetModules();
});

test('Happy path without interims', async () => {
  const recognizedResult = {
    json: JSON.stringify({
      RecognitionStatus: 'Success',
      Offset: 1800000,
      Duration: 48100000,
      NBest: [{
        Confidence: 0.2331869,
        Lexcial: 'no',
        ITN: 'no',
        MaskedITN: 'no',
        Display: 'No.'
      }]
    }),
    reason: 3
  };

  jest.setMock('../SpeechSDK', ({
    ...MOCK_SPEECH_SDK,
    SpeechRecognizer: class extends MOCK_SPEECH_SDK.SpeechRecognizer {
      recognizeOnceAsync(success) {
        setImmediate(() => this.recognized(this, { result: recognizedResult }));
        setImmediate(() => success(recognizedResult));
      }
    }
  }));

  const { default: createSpeechRecognitionPonyfill } = require('./createSpeechRecognitionPonyfill');
  const { SpeechRecognition } = createSpeechRecognitionPonyfill({
    region: 'westus',
    subscriptionKey: 'SUBSCRIPTION_KEY'
  });

  const speechRecognition = new SpeechRecognition();
  const getEvents = captureSpeechEvents(speechRecognition);

  await new Promise(resolve => {
    speechRecognition.addEventListener('end', resolve);
    speechRecognition.start();
    jest.runAllImmediates();
  });

  expect(getEvents()).toMatchSnapshot();
});
