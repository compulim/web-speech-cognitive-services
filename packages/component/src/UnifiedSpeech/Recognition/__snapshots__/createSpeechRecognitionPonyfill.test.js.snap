// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Happy path with 2 interims 1`] = `
Array [
  Object {
    "type": "start",
  },
  Object {
    "type": "audiostart",
  },
  Object {
    "type": "soundstart",
  },
  Object {
    "type": "speechstart",
  },
  Object {
    "results": Array [
      Array [
        Object {
          "confidence": 0.5,
          "transcript": "hello",
        },
      ],
    ],
    "type": "result",
  },
  Object {
    "results": Array [
      Array [
        Object {
          "confidence": 0.5,
          "transcript": "john",
        },
      ],
    ],
    "type": "result",
  },
  Object {
    "type": "speechend",
  },
  Object {
    "type": "soundend",
  },
  Object {
    "type": "audioend",
  },
  Object {
    "results": Array [
      Array [
        Object {
          "confidence": 0.9,
          "transcript": "Hello, John.",
        },
      ],
    ],
    "type": "result",
  },
  Object {
    "type": "end",
  },
]
`;

exports[`Happy path without interims 1`] = `
Array [
  Object {
    "type": "start",
  },
  Object {
    "type": "audiostart",
  },
  Object {
    "type": "soundstart",
  },
  Object {
    "type": "speechstart",
  },
  Object {
    "type": "speechend",
  },
  Object {
    "type": "soundend",
  },
  Object {
    "type": "audioend",
  },
  Object {
    "results": Array [
      Array [
        Object {
          "confidence": 0.2331869,
          "transcript": "No.",
        },
      ],
    ],
    "type": "result",
  },
  Object {
    "type": "end",
  },
]
`;

exports[`Muted microphone 1`] = `
Array [
  Object {
    "type": "start",
  },
  Object {
    "type": "audiostart",
  },
  Object {
    "type": "audioend",
  },
  Object {
    "error": "no-speech",
    "type": "error",
  },
  Object {
    "type": "end",
  },
]
`;

exports[`Network error before start 1`] = `
Array [
  Object {
    "type": "start",
  },
  Object {
    "type": "audiostart",
  },
  Object {
    "type": "audioend",
  },
  Object {
    "error": "network",
    "type": "error",
  },
  Object {
    "type": "end",
  },
]
`;

exports[`Microphone blocked 1`] = `
Array [
  Object {
    "error": "not-allowed",
    "type": "error",
  },
]
`;
